File: task1.py
import random

def t1():
    N = int(input("Введите размерность матрицы N: "))  

    A = [[random.randint(-10, 10) for _ in range(N)] for _ in range(N)]


    print("Исходная матрица A:")
    for row in A:
        print(row)

    main_diag = [A[i][i] for i in range(N)]
    secondary_diag = [A[i][N - 1 - i] for i in range(N)]

    B = [main_diag, secondary_diag]


    print("\nНовая матрица B (диагонали):")
    for row in B:
        print(row)

def t2():
   M, N = int(input("Введите количество строк M: ")), int(input("Введите количество столбцов N: "))

# Генерация матрицы случайными числами от -20 до 20
A = [[random.randint(-20, 20) for _ in range(N)] for _ in range(M)]

print("Исходная матрица:")
for row in A:
    print(row)

# Сбор всех элементов в один список для поиска max и second_min
all_elements = [elem for row in A for elem in row]

# Поиск максимума и второго наименьшего
max_elem = max(all_elements)
sorted_unique = sorted(set(all_elements))
second_min = sorted_unique[1] if len(sorted_unique) > 1 else sorted_unique[0]  # если все числа одинаковые

replacement_value = max_elem - second_min

# Замена диагональных элементов
for i in range(M):
    for j in range(N):
        if i == j or i + j == N - 1:
            A[i][j] = replacement_value

print("\nМатрица после замены диагональных элементов:")
for row in A:
    print(row)
    
def is_valid_sudoku(board):
    rows = [set() for _ in range(6)]
    cols = [set() for _ in range(6)]
    boxes = [set() for _ in range(4)]

    for r in range(6):
        for c in range(6):
            num = board[r][c]
            if num == '.':
                continue
            if num in rows[r] or num in cols[c]:
                return False
            box_index = (r // 3) * 2 + (c // 3)
            if num in boxes[box_index]:
                return False
            rows[r].add(num)
            cols[c].add(num)
            boxes[box_index].add(num)
    return True

def input_board():
    board = [['.' for _ in range(6)] for _ in range(6)]
    print("Введите заполненные ячейки в формате: row col value (1-6). Введите 'end' для завершения.")
    while True:
        line = input(">> ")
        if line.lower() == 'end':
            break
        try:
            r, c, v = line.strip().split()
            r, c = int(r)-1, int(c)-1
            if not (0 <= r < 6 and 0 <= c < 6 and v in '123456789'):
                print("Ошибка координат или значения.")
                continue
            board[r][c] = v
        except:
            print("Ошибка ввода. Пример: 1 2 5")
    return board

def generate_random_board(filled_cells=10):
    board = [['.' for _ in range(6)] for _ in range(6)]
    attempts = 0
    while filled_cells > 0 and attempts < 100:
        r = random.randint(0, 5)
        c = random.randint(0, 5)
        v = str(random.randint(1, 6))
        if board[r][c] != '.':
            continue
        board[r][c] = v
        if not is_valid_sudoku(board):
            board[r][c] = '.'
        else:
            filled_cells -= 1
        attempts += 1
    return board

def print_board(board):
    print("\nТекущая доска:")
    for row in board:
        print(" ".join(row))
    print()

def t3():
    print("Выберите режим:")
    print("1 — Ввод вручную")
    print("2 — Случайная генерация")
    choice = input(">> ")
    if choice == '1':
        board = input_board()
    else:
        board = generate_random_board()

    print_board(board)

    if is_valid_sudoku(board):
        print("Доска корректна по правилам судоку.")
    else:
        print("Нарушены правила судоку.")
        
        
def main():
    print("Выберите задачу:")
    print("1. Задача 1")
    print("2. Задача 2")
    print("3. Задача 3")
    
    choice = input("Введите номер задачи (1-3): ")
    
    if choice == '1':
        t1()
    elif choice == '2':
        t2()
    elif choice == '3':
        t3()
    else:
        print("Неверный выбор. Пожалуйста, выберите 1, 2 или 3.")
        
main()

File: task1.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <set>
#include <unordered_set>

using namespace std;

void t1() {
    int N;
    cout << "Введите размерность матрицы N (N x N): ";
    cin >> N;
    vector<vector<int>> A(N, vector<int>(N));
    vector<vector<int>> B(2, vector<int>(N));

    srand(time(0)); // Инициализация генератора случайных чисел

    cout << "Исходная матрица A:\n";
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            A[i][j] = rand() % 21 - 10; 
            cout << A[i][j] << "\t";
        }
        cout << "\n";
    }

    // Заполнение матрицы B диагоналями
    for (int i = 0; i < N; ++i) {
        B[0][i] = A[i][i];         // главная диагональ
        B[1][i] = A[i][N - 1 - i]; // побочная диагональ
    }

    // Вывод матрицы B
    cout << "\nМатрица B (две строки с диагоналями):\n";
    for (int i = 0; i < 2; ++i) {
        for (int j = 0; j < N; ++j) {
            cout << B[i][j] << "\t";
        }
        cout << "\n";
    }
}

void t2() {
    const int M = 5;
    const int N = 6;
    vector<vector<int>> A(M, vector<int>(N));
    vector<int> all_elements;

    srand(time(0));

    // Генерация матрицы и сбор всех элементов
    cout << "Исходная матрица:\n";
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            A[i][j] = rand() % 41 - 20; // от -20 до 20
            cout << A[i][j] << "\t";
            all_elements.push_back(A[i][j]);
        }
        cout << "\n";
    }

    // Нахождение максимума и второго минимального
    int max_elem = *max_element(all_elements.begin(), all_elements.end());

    // Используем set для получения уникальных значений и сортировки
    set<int> unique_elements(all_elements.begin(), all_elements.end());
    int second_min = *unique_elements.begin();
    if (unique_elements.size() > 1) {
        auto it = unique_elements.begin();
        ++it;
        second_min = *it;
    }

    int replacement_value = max_elem - second_min;

    // Замена элементов на диагоналях
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            if (i == j || i + j == N - 1) {
                A[i][j] = replacement_value;
            }
        }
    }

    // Вывод результата
    cout << "\nМатрица после замены диагональных элементов:\n";
    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            cout << A[i][j] << "\t";
        }
        cout << "\n";
    }
}

bool isValidSudoku(const vector<vector<char>>& board) {
    vector<unordered_set<char>> rows(6), cols(6), boxes(4);
    for (int r = 0; r < 6; ++r) {
        for (int c = 0; c < 6; ++c) {
            char num = board[r][c];
            if (num == '.') continue;
            int boxIndex = (r / 3) * 2 + (c / 3);
            if (rows[r].count(num) || cols[c].count(num) || boxes[boxIndex].count(num))
                return false;
            rows[r].insert(num);
            cols[c].insert(num);
            boxes[boxIndex].insert(num);
        }
    }
    return true;
}

vector<vector<char>> inputBoard() {
    vector<vector<char>> board(6, vector<char>(6, '.'));
    cout << "Введите заполненные ячейки (строка столбец значение), от 1 до 6. Введите -1 для завершения:\n";
    while (true) {
        int r, c;
        char val;
        cout << ">> ";
        cin >> r;
        if (r == -1) break;
        cin >> c >> val;
        if (r < 1 || r > 6 || c < 1 || c > 6 || val < '1' || val > '9') {
            cout << "Неверный ввод. Повторите.\n";
            continue;
        }
        board[r - 1][c - 1] = val;
    }
    return board;
}

vector<vector<char>> generateRandomBoard(int filled = 10) {
    vector<vector<char>> board(6, vector<char>(6, '.'));
    srand(time(0));
    int attempts = 0;
    while (filled > 0 && attempts < 100) {
        int r = rand() % 6;
        int c = rand() % 6;
        char val = '1' + rand() % 6;
        if (board[r][c] != '.') continue;
        board[r][c] = val;
        if (!isValidSudoku(board)) {
            board[r][c] = '.';
        } else {
            --filled;
        }
        ++attempts;
    }
    return board;
}

void printBoard(const vector<vector<char>>& board) {
    cout << "\nТекущая доска:\n";
    for (const auto& row : board) {
        for (char c : row) cout << c << ' ';
        cout << '\n';
    }
}

void t3() {
    cout << "Выберите режим:\n1 — Ввод вручную\n2 — Случайная генерация\n>> ";
    int choice;
    cin >> choice;
    vector<vector<char>> board;
    if (choice == 1)
        board = inputBoard();
    else
        board = generateRandomBoard();

    printBoard(board);

    if (isValidSudoku(board))
        cout << "Доска корректна по правилам судоку.\n";
    else
        cout << "Нарушены правила судоку.\n";
}

int main() {
    cout << "Выберите задачу:\n1 — Задача 1\n2 — Задача 2\n3 — Задача 3\n>> ";
    int task;
    cin >> task;
    switch (task) {
        case 1: t1(); break;
        case 2: t2(); break;
        case 3: t3(); break;
        default: cout << "Неверный выбор задачи.\n"; break;
    }
    return 0;
}